File: script/python/game/scene/game_player.py
# -*- encoding:utf-8 -*-
import Scene
from proto.pb_message import Message
from game.util import logger
import game.util.db_util
from game.util.id_mgr import IDMgr

from game.db.tbl.tbl_player import TblPlayer
import game.db.tbl.tbl_item

import game.scene.player.item_mgr
import game.scene.player.msg_handler

import game.scene.game_actor
from game.service.service_addr import LOCAL_DB_SERVICE_ADDR
from game.service.service_addr import LOCAL_SCENE_CTRL_SERVICE_ADDR
from game.service.service_addr import LOCAL_SERVICE_GROUP
from game.service.service_addr import ServiceAddr
from game.util.const import ErrorCode
from game.util.multi_index_container import MultiIndexElement


class GamePlayer(game.scene.game_actor.GameActor, MultiIndexElement):
    # _c_cmd = game.util.cmd_util.CmdDispatch("c_player")

    ATTR_ROLE_ID = "role_id"
    ATTR_CONN_ID = "conn_id"

    @staticmethod
    def define_multi_index_attr_names():
        return GamePlayer.ATTR_ROLE_ID, GamePlayer.ATTR_CONN_ID

    def __init__(self, e_player, game_scene, conn_id, role_id):
        MultiIndexElement.__init__(self)
        game.scene.game_actor.GameActor.__init__(self, e_player[1])
        self.native_obj = Scene.Player(e_player, self)
        # self.actor_id = e_player[1]
        self.game_scene = game_scene
        self.conn_id = conn_id
        self.role_id = role_id
        self.server_id = IDMgr.get_server_id_by_uid(role_id)
        self._init_mgr()

    def _init_mgr(self):
        self._item_mgr = game.scene.player.item_mgr.ItemMgr(self)
        self.msg_handler = game.scene.player.msg_handler.MsgHandler(self)

    def init_player_data(self, tables):
        # print(tables)
        tb_player = tables["player"][0]
        self.name = tb_player.role_name
        self.account = tb_player.account

        self._item_mgr.on_load_item(tables.get("item"))

    def is_player(self):
        return True

    def send_msg_to_client(self, msg):
        self.game_scene.service.send_msg_to_client(self.conn_id, msg)

    def send_msg_to_client_kcp(self, msg):
        self.game_scene.service.send_msg_to_client_kcp(self.conn_id, msg)

    def send_msg_to_service(self, dst_srv, msg):
        self.game_scene.service.send_msg_to_service(dst_srv, msg)

    def pack_born_info(self, msg):
        player_info = msg.player_list.add()
        player_info.actor_id = self.actor_id
        player_info.name = self.name

    def on_actor_enter_sight(self, actor):
        if not self.check_can_see(actor):
            return
        msg = Message.create_msg_by_id(Message.MSG_ID_ACTOR_BORN)
        actor.pack_born_info(msg)
        self.send_msg_to_client(msg)

    def on_actor_leave_sight(self, actor):
        msg = Message.create_msg_by_id(Message.MSG_ID_ACTOR_DISSOLVE)
        msg.actor_ids.append(actor.actor_id)
        self.send_msg_to_client(msg)
        # print("on_actor_leave_sight")

    def on_enter_scene(self):
        game_scene = self.game_scene
        game_scene.service.rpc_call(LOCAL_SCENE_CTRL_SERVICE_ADDR, "Player_EnterScene", role_id=self.role_id,
                                    scene_uid=game_scene.scene_uid, scene_id=game_scene.scene_id)

    def on_leave_scene(self):
        pass

    def on_leave_game(self):
        self.game_scene.service.rpc_call(LOCAL_SCENE_CTRL_SERVICE_ADDR, "Player_LeaveGame", role_id=self.role_id)

    def try_switch_scene(self, server_id, scene_id):
        if LOCAL_SERVICE_GROUP == server_id and self.game_scene.scene_id == scene_id:
            logger.log_info("try switch scene fail, cur scene id={0}, role_id:{1}", scene_id, self.role_id)
            return

        if LOCAL_SERVICE_GROUP == server_id:
            self._local_switch_scene(server_id, scene_id)
        else:
            self._remote_switch_scene(server_id, scene_id)

    def _local_switch_scene(self, server_id, scene_id):
        def _on_finish(err_code, scene_uid):
            if err_code != ErrorCode.OK:
                logger.log_error("switch scene failed, role_id:%d, server_id:%d, scene_id:%d, err:%d", self.role_id, server_id,
                                scene_id, err_code)
                return
            self.switch_local_scene(scene_uid)

        scene_ctrl_addr = ServiceAddr.make_scene_ctrl_addr(server_id)
        future = self.game_scene.service.rpc_call(scene_ctrl_addr, "Player_SwitchSceneReq", -1, scene_id=scene_id, role_id=self.role_id)
        future.on_fin += _on_finish
        future.on_timeout += _on_finish

    def _remote_switch_scene(self, server_id, scene_id):
        def _on_finish(err_code, server_id=None, scene_uid=None, gateway_addr=None, token=None):
            if err_code != ErrorCode.OK:
                logger.log_error("switch remote scene failed, role_id:{}, server_id:{}, scene_id:{}, err:{}", self.role_id,
                                server_id, scene_id, err_code)
                return
            self.switch_remote_scene(server_id, scene_uid, gateway_addr, token)

        scene_ctrl_addr = ServiceAddr.make_scene_ctrl_addr(server_id)
        future = self.game_scene.service.rpc_call(scene_ctrl_addr, "Player_RemoteSwitchSceneReq", -1, scene_id=scene_id, role_id=self.role_id)
        future.on_fin += _on_finish
        future.on_timeout += _on_finish

    def switch_remote_scene(self, server_id, scene_uid, gateway_addr, token):
        logger.log_info("switch remote scene, role_id:{}, server_id:{}, scene_uid:{}, gateway_addr:{}",
                       self.role_id, server_id, scene_uid, gateway_addr)
        scene_ctrl_addr = ServiceAddr.make_scene_ctrl_addr(self.server_id)
        self.game_scene.service.rpc_call(scene_ctrl_addr, "Player_SwitchToRemoteScene", -1,
                                         role_id=self.role_id, server_id=server_id, scene_uid=scene_uid)
        msg = Message.create_msg_by_id(Message.MSG_ID_SWITCH_REMOTE_SERVER)
        msg.remote_ip = gateway_addr["ip"]
        msg.remote_port = gateway_addr["port"]
        msg.remote_udp_port = gateway_addr["udp_port"]
        msg.token = token
        self.send_msg_to_client(msg)

    def switch_local_scene(self, scene_uid):
        logger.log_info("switch local scene, role_id:{}, scene_uid:{}", self.role_id, scene_uid)
        self.game_scene.service.rpc_call(LOCAL_SCENE_CTRL_SERVICE_ADDR, "Player_SwitchScene", -1,
                                         conn_id=self.conn_id, role_id=self.role_id, scene_uid=scene_uid)
        self.leave_scene("switch_scene")

    def leave_scene(self, reason):
        self.game_scene.remove_player(self.role_id, reason)

File: script/lua/game/scene/game_player.lua
require("game.scene.scene_entity")
require("util.logger")
require("game.scene.player.item_mgr")
require("game.scene.player.msg_handler")
require("base.service_type")

clsGamePlayer = clsSceneEntity:Inherit("clsGamePlayer")

clsGamePlayer.index_roleId = "role_id"
clsGamePlayer.index_connId = "conn_id"

function clsGamePlayer:__init__(game_scene, engine_obj, role_id, name)
    Super(clsGamePlayer).__init__(self, game_scene, engine_obj:getEntityId(), {clsGamePlayer.index_roleId, clsGamePlayer.index_connId})
    self.engineObj = engine_obj
    self.role_id = role_id
    self.server_id = IDMgr.get_server_id_by_uid(role_id)
    self.name = name
    self.conn_id = engine_obj:getConnId()
    self:_init_mgr()
end

function clsGamePlayer:_init_mgr()
    self._item_mgr = clsItemMgr:New(self)
    self.msg_handler = clsMsgHandler:New(self)
end

function clsGamePlayer:init_player_data(tbls)
    -- # print(tables)
    local tb_player = tbls.player.rows[1]
    self.name = tb_player.role_name
    self.account = tb_player.account

    self._item_mgr:on_load_item(tbls.item)
end

function clsGamePlayer:is_player()
    return true
end

function clsGamePlayer:send_msg_to_client(msgId, msg)
    -- self.game_scene.service:sendMsgToClient(self.conn_id, msgId, msg)
    local data = encodeMsg(msgId, msg)
    self.engineObj:sendToClient(msgId, data, string.len(data))
end

function clsGamePlayer:send_msg_to_sight(msgId, msg)
    -- self.game_scene.service:sendMsgToClient(self.conn_id, msgId, msg)
    local data = encodeMsg(msgId, msg)
    self.engineObj:sendToSight(msgId, data, string.len(data))
end

function clsGamePlayer:send_msg_to_client_kcp(msgId, msg)
    self.game_scene.service:sendMsgToClientKCP(self.conn_id, msgId, msg)
end

function clsGamePlayer:send_msg_to_service(dst_srv, msgId, msg)
    self.game_scene.service:sendMsgToService(dst_srv, msgId, msg)
end

function clsGamePlayer:pack_born_info(msg)
    msg.player_list = msg.player_list or {}
    local player_info = {}
    player_info.actor_id = self.entity_id
    player_info.name = self.name
    table.insert(msg.player_list, player_info)
end

function clsGamePlayer:on_entity_enter_sight(entity)
    if not self:check_can_see(entity) then return end
    local msg = {}
    entity:pack_born_info(msg)
    self:send_msg_to_client(MSG_ID_ACTOR_BORN, msg)
end

function clsGamePlayer:on_entity_leave_sight(entity)
    local msg = {actor_ids = {}}
    table.insert(msg.actor_ids, entity.entity_id)
    self:send_msg_to_client(MSG_ID_ACTOR_DISSOLVE, msg)
    print("on_entity_leave_sight")
end

function clsGamePlayer:on_enter_scene()
    local game_scene = self.game_scene
    game_scene.service:callRpc(LOCAL_SERVICE_SCENE_CTRL, "Player_EnterScene", -1, {role_id=self.role_id,
                                scene_uid=game_scene.scene_uid, scene_id=game_scene.scene_id})
end

function clsGamePlayer:on_reconnect(conn_id)
    self:change_conn_id(conn_id)
    logger.logInfo("reconnect, role_id:%d, conn_id:%d", self.role_id, conn_id)
end

function clsGamePlayer:change_conn_id(conn_id)
    self:change_index_attr_value(clsGamePlayer.index_connId, conn_id)
    self.conn_id = conn_id
    self.engineObj:setConnId(conn_id)
end

function clsGamePlayer:on_leave_scene()
end

function clsGamePlayer:on_leave_game()
    self.game_scene.service:callRpc(LOCAL_SERVICE_SCENE_CTRL, "Player_LeaveGame", -1, {role_id=self.role_id})
end

function clsGamePlayer:try_switch_scene(server_id, scene_id)
    if SERVER_GROUP_ID == server_id and self.game_scene.scene_id == scene_id then
        logger.logInfo("try switch scene fail, cur scene id=%d, role_id:%d", scene_id, self.role_id)
        return
    end
    local role_id = self.role_id

    -- local function _on_finish(err_code, scene_uid)
    --     if err_code ~= ErrorCode.OK then
    --         logger.logError("switch scene failed, role_id:%d, server_id:%d, scene_id:%d, err:%d", role_id, server_id, scene_id, err_code)
    --         return
    --     end
    --     self:switch_scene(server_id, scene_uid)
    -- end

    -- local scene_ctrl_addr = make_scene_ctrl_addr(server_id)

    if SERVER_GROUP_ID == server_id then
        self:_local_switch_scene(server_id, scene_id)
        -- local future = self.game_scene.service:callRpc(scene_ctrl_addr, "Player_SwitchSceneReq", -1, {scene_id=scene_id, role_id=role_id})
        -- future:regCallback(_on_finish)
    else
        self:_remote_switch_scene(server_id, scene_id)
        -- local future = self.game_scene.service:callRpc(scene_ctrl_addr, "Player_RemoteSwitchSceneReq", -1, {scene_id=scene_id, role_id=role_id})
        -- future:regCallback(_on_finish)
    end
end

-- 在本服切换场景
function clsGamePlayer:_local_switch_scene(server_id, scene_id)
    local function _on_finish(err_code, scene_uid)
        if err_code ~= ErrorCode.OK then
            logger.logError("switch scene failed, role_id:%d, server_id:%d, scene_id:%d, err:%d", role_id, server_id, scene_id, err_code)
            return
        end
        self:switch_local_scene(server_id, scene_uid)
    end

    local scene_ctrl_addr = make_scene_ctrl_addr(server_id)
    local future = self.game_scene.service:callRpc(scene_ctrl_addr, "Player_SwitchSceneReq", -1, {scene_id=scene_id, role_id=self.role_id})
    future:regCallback(_on_finish)
end

-- 切换到其他服场景
function clsGamePlayer:_remote_switch_scene(server_id, scene_id)
    local function _on_finish(err_code, result)
        if err_code ~= ErrorCode.OK then
            logger.logError("switch remote scene failed, role_id:%d, server_id:%d, scene_id:%d, err:%d", self.role_id, server_id, scene_id, err_code)
            return
        end
        local server_id = result.server_id
        local scene_uid = result.scene_uid
        local gateway_addr = result.gateway_addr
        local token = result.token
        self:switch_remote_scene(server_id, scene_uid, gateway_addr, token)
    end

    local scene_ctrl_addr = make_scene_ctrl_addr(server_id)
    local future = self.game_scene.service:callRpc(scene_ctrl_addr, "Player_RemoteSwitchSceneReq", -1, {scene_id=scene_id, role_id=self.role_id})
    future:regCallback(_on_finish)
end

function clsGamePlayer:is_local_server()
    return self.server_id == SERVER_GROUP_ID
end

function clsGamePlayer:switch_local_scene(server_id, scene_uid)
    logger.logInfo("switch scene, role_id:%d, server_id:%d, scene_uid:%d", self.role_id, server_id, scene_uid)
    self.game_scene.service:callRpc(LOCAL_SERVICE_SCENE_CTRL, "Player_SwitchScene", -1,
                                        {conn_id=self.conn_id, role_id=self.role_id, scene_uid=scene_uid})
    self:leave_scene("switch_scene")
end

function clsGamePlayer:switch_remote_scene(server_id, scene_uid, gateway_addr, token)
    logger.logInfo("switch remote scene, role_id:%d, server_id:%d, scene_uid:%d, gateway_addr:%s", 
                        self.role_id, server_id, scene_uid, StrUtil.tableToStr(gateway_addr))
    local scene_ctrl_addr = make_scene_ctrl_addr(self.server_id)
    self.game_scene.service:callRpc(scene_ctrl_addr, "Player_SwitchToRemoteScene", -1,
                                        {role_id=self.role_id, server_id=server_id, scene_uid=scene_uid})
    local msg = {}
    msg.remote_ip = gateway_addr.ip
    msg.remote_port = gateway_addr.port
    msg.remote_udp_port = gateway_addr.udp_port
    msg.token = token
    self:send_msg_to_client(MSG_ID_SWITCH_REMOTE_SERVER, msg)
    -- self:leave_scene("switch_remote_scene")
    -- self.game_scene:tick_player(self.role_id, "switch_remote_scene")
end

function clsGamePlayer:leave_scene(reason)
    self.game_scene:remove_player(self.role_id, reason)
end

function clsGamePlayer:sendGMCmdRsp(cmd, result)
    local rsp_msg = {}
    rsp_msg.cmd = cmd
    rsp_msg.msg = result or ""
    self:send_msg_to_client(MSG_ID_GM_CMD_RSP, rsp_msg)
end